test/unittest/statistics/rtps/RTPSStatisticsTests.cpp

```diff
@@ -0,0 +1,225 @@
+        EXPECT_FALSE(writer->remove_statistics_listener(listener2));
+        // + succeeds to remove a known listener
+        EXPECT_TRUE(writer->remove_statistics_listener(listener1));
+        // + fails if a listener is already removed
+        EXPECT_FALSE(writer->remove_statistics_listener(listener1));
+
+        // test the reader apis
+        // + fails if no listener has been yet added
+        EXPECT_FALSE(reader->remove_statistics_listener(listener1));
+        // + fails to add an empty listener
+        EXPECT_FALSE(reader->add_statistics_listener(nolistener));
+        // + succeeds to add a new listener
+        ASSERT_TRUE(reader->add_statistics_listener(listener1));
+        // + fails to add multiple times the same listener
+        EXPECT_FALSE(reader->add_statistics_listener(listener1));
+        // + fails if an unknown listener is removed
+        EXPECT_FALSE(reader->remove_statistics_listener(listener2));
+        // + succeeds to remove a known listener
+        EXPECT_TRUE(reader->remove_statistics_listener(listener1));
+        // + fails if a listener is already removed
+        EXPECT_FALSE(reader->remove_statistics_listener(listener1));
+    }
+
+    // Check PUBLICATION_THROUGHPUT and SUBSCRIPTION_THROUGHPUT callbacks are performed
+    {
+        using namespace ::testing;
+
+        auto participant_listener = make_shared<MockListener>();
+        auto writer_listener = make_shared<MockListener>();
+        auto reader_listener = make_shared<MockListener>();
+        ASSERT_TRUE(participant->add_statistics_listener(participant_listener,
+                fastdds::statistics::EventKind::DISCOVERED_ENTITY));
+        ASSERT_TRUE(writer->add_statistics_listener(writer_listener));
+        ASSERT_TRUE(reader->add_statistics_listener(reader_listener));
+
+        // We must received the discovery time of each endpoint
+        EXPECT_CALL(*writer_listener, on_statistics_data)
+                .Times(2);
+
+        // match writer and reader on a dummy topic
+        TopicAttributes Tatt;
+        Tatt.topicKind = NO_KEY;
+        Tatt.topicDataType = "string";
+        Tatt.topicName = "statisticsTopic";
+        WriterQos Wqos;
+        ReaderQos Rqos;
+        participant->registerWriter(writer, Tatt, Wqos);
+        participant->registerReader(reader, Tatt, Rqos);
+
+        // Check callbacks on data exchange, at least, we must received:
+        // + RTPSWriter: PUBLICATION_THROUGHPUT, RTPS_SENT, RESENT_DATAS,
+        //               GAP_COUNT, DATA_COUNT, SAMPLE_DATAS & PHYSICAL_DATA
+        //   optionally: ACKNACK_COUNT & NACKFRAG_COUNT
+        EXPECT_CALL(*writer_listener, on_statistics_data)
+                .Times(AtMost(7));
+
+        // + RTPSReader: SUBSCRIPTION_THROUGHPUT, RTPS_LOST, DATA_COUNT,
+        //               SAMPLE_DATAS & PHYSICAL_DATA
+        //   optionally: HEARTBEAT_COUNT
+        EXPECT_CALL(*reader_listener, on_statistics_data)
+                .Times(AtMost(5));
```

Comment:
I do not think this is enough to test the callback reception. This test probably needs a SetUp which enables the collection and publication of the statistics data. This depends on the implementation to be done.

Reply:
That's a dds layer feature. It was discussed in the design that statistics in the rtps layer was enabled by mere listener assignment.